spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/security/oauth2/client/OAuth2ClientAutoConfiguration.java

```diff
@@ -0,0 +1,182 @@
+import org.springframework.boot.context.properties.bind.Bindable;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.ConditionContext;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.ConfigurationCondition;
+import org.springframework.core.annotation.Order;
+import org.springframework.core.env.Environment;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+import org.springframework.security.config.annotation.web.builders.HttpSecurity;
+import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration;
+import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
+import org.springframework.security.config.annotation.web.configurers.oauth2.client.OAuth2LoginConfigurer;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationProperties;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.client.registration.InMemoryClientRegistrationRepository;
+
+/**
+ * {@link EnableAutoConfiguration Auto-configuration} for Spring Security OAuth 2.0 / OpenID Connect 1.0 client support.
+ *
+ * @author Joe Grandja
+ * @since 2.0.0
+ */
+@Configuration
+@ConditionalOnWebApplication(type = Type.SERVLET)
+@ConditionalOnClass({EnableWebSecurity.class, ClientRegistration.class})
+@AutoConfigureBefore(SecurityAutoConfiguration.class)
+public class OAuth2ClientAutoConfiguration {
+
+	private static final String CLIENT_PROPERTY_PREFIX = "security.oauth2.client";
+
+	private static final String CLIENT_ID_PROPERTY = "client-id";
+
+	private static final String USER_INFO_URI_PROPERTY = "user-info-uri";
+
+	private static final String USER_NAME_ATTR_NAME_PROPERTY = "user-name-attribute-name";
+
+	static Set<String> getClientKeys(Environment environment) {
+		return getClientPropertiesByClient(environment).keySet();
+	}
+
+	static Map<String, Map> getClientPropertiesByClient(Environment environment) {
+		Map<String, Object> clientPropertiesByClient = Binder.get(environment)
+				.bind(CLIENT_PROPERTY_PREFIX, Bindable.mapOf(String.class, Object.class))
+				.orElse(new HashMap<>());
+
+		// Filter out clients that don't have the client-id property set
+		return clientPropertiesByClient.entrySet().stream()
+				.map(e -> new AbstractMap.SimpleImmutableEntry<>(e.getKey(), (Map) e.getValue()))
+				.filter(e -> e.getValue().containsKey(CLIENT_ID_PROPERTY))
+				.collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+	}
+
+	@Order(1)
```

Comment:
Why is this ordered? It looks like it's to ensure that it is processed before `OAuth2LoginConfiguration`. I am not sure that it will be guaranteed to have that effect when they're inner-classes rather than classes registered directly with the context. Our preferred approach for this is to use `@Import` where the order of the referenced classes is guaranteed to be honoured.

Reply:
I've split-up `OAuth2LoginConfiguration` into `ClientRegistrationRepositoryAutoConfiguration` and  `OAuth2LoginAutoConfiguration` and leveraged `@AutoConfigureBefore` to ensure the correct ordering. Will push those changes shortly.